### PR TITLE
OCPBUGS-43095: add duplicate_te_header_total metric

### PR DESCRIPTION
**Pull request by frobware**
This PR address an issue encountered when upgrading from HAProxy 2.2 in OpenShift 4.13 to HAProxy 2.6 in OpenShift 4.14, where responses are rejected due to adherence to RFC 7230 concerning duplicate Transfer-Encoding headers.

HAProxy 2.2 does not impose this restriction, but post-upgrade, users face rejected responses when multiple Transfer-Encoding headers are present, violating RFC 7230 specifications. HAProxy rejects the connection with a 502 Bad Gateway status code.

To mitigate this issue, we are modifying HAProxy 2.2 to include functionality that detects duplicate Transfer-Encoding headers per response. This enhancement allows customers to utilise a new metric, `duplicate_te_header_total`, which shows routes that currently record duplicate Transfer-Encoding headers, enabling proactive identification of affected routes under OpenShift 4.13.

Note that the changes in HAProxy 2.6 that enforce these restrictions are detailed in the following commits:

- [Commit 545fbba](https://github.com/haproxy/haproxy/commit/545fbba2733a1f611ce940ecd5fbea1130b6d503): Restricts duplicate Transfer-Encoding headers.
- [Commit da3adeb](https://github.com/haproxy/haproxy/commit/da3adebd06a85cacbdca577f03e433243c351fe9): Introduces rejection of transfer-encoding values other than "chunked".

Related PR for e2e testing: https://github.com/openshift/cluster-ingress-operator/pull/1150.

### CODE REVIEW

**Comment by frobware** [comment]
HAProxy 2.2.24 patch:

```diff
From 0b3577ce1e0236498756fba65ee0ef9d436f8a10 Mon Sep 17 00:00:00 2001
From: Andrew McDermott <amcdermo@redhat.com>
Date: Thu, 19 Sep 2024 10:56:53 +0100
Subject: UPSTREAM: <carry>: Record duplicate Transfer-Encoding headers

Add support for tracking servers responding with duplicate
Transfer-Encoding headers within a single response. Increment the
counter for the relevant backend when duplicates are detected. Introduce
the new te_dup stat to enable interrogation of backends experiencing
this issue through the stats socket.
---
 include/haproxy/counters-t.h | 2 ++
 include/haproxy/h1.h         | 1 +
 include/haproxy/stats-t.h    | 1 +
 src/h1.c                     | 5 +++++
 src/mux_h1.c                 | 4 ++++
 src/stats.c                  | 3 +++
 6 files changed, 16 insertions(+)

diff --git a/include/haproxy/counters-t.h b/include/haproxy/counters-t.h
index b896c4d36..05d2f69aa 100644
--- a/include/haproxy/counters-t.h
+++ b/include/haproxy/counters-t.h
@@ -86,6 +86,8 @@ struct be_counters {
 	long long denied_req;                   /* blocked requests because of security concerns */
 	long long denied_resp;                  /* blocked responses because of security concerns */
 
+	long long duplicate_te_header;          /* number of times responses included duplicated Transfer-Encoding header */
+
 	long long connect;                      /* number of connection establishment attempts */
 	long long reuse;                        /* number of connection reuses */
 	long long failed_conns;                 /* failed connect() attempts (BE only) */
diff --git a/include/haproxy/h1.h b/include/haproxy/h1.h
index 8835f37f6..f1284f53b 100644
--- a/include/haproxy/h1.h
+++ b/include/haproxy/h1.h
@@ -95,6 +95,7 @@ enum h1m_state {
 #define H1_MF_CLEAN_CONN_HDR    0x00001000 // skip close/keep-alive values of connection headers during parsing
 #define H1_MF_METH_CONNECT      0x00002000 // Set for a response to a CONNECT request
 #define H1_MF_METH_HEAD         0x00004000 // Set for a response to a HEAD request
+#define H1_MF_DUPLICATE_TE      0x00008000 // Set when a duplicated Transfer-Encoding header is detected
 
 /* Note: for a connection to be persistent, we need this for the request :
  *   - one of CLEN or CHNK
diff --git a/include/haproxy/stats-t.h b/include/haproxy/stats-t.h
index 300454a14..caa318c09 100644
--- a/include/haproxy/stats-t.h
+++ b/include/haproxy/stats-t.h
@@ -429,6 +429,7 @@ enum stat_field {
 	ST_F_SAFE_CONN_CUR,
 	ST_F_USED_CONN_CUR,
 	ST_F_NEED_CONN_EST,
+	ST_F_TE_DUP,
 
 	/* must always be the last one */
 	ST_F_TOTAL_FIELDS
diff --git a/src/h1.c b/src/h1.c
index bb8acfb34..07b92fb4c 100644
--- a/src/h1.c
+++ b/src/h1.c
@@ -277,6 +277,7 @@ int h1_headers_to_hdr_list(char *start, const char *stop,
 	unsigned int col;  /* position of the colon */
 	unsigned int eol;  /* end of line */
 	unsigned int sov;  /* start of value */
+	unsigned int te_header_seen = 0; /* Tracks if a Transfer-Encoding header has been encountered. */
 	union h1_sl sl;
 	int skip_update;
 	int restarting;
@@ -807,6 +808,10 @@ int h1_headers_to_hdr_list(char *start, const char *stop,
 
 				if (isteqi(n, ist("transfer-encoding"))) {
 					h1_parse_xfer_enc_header(h1m, v);
+					if (te_header_seen && (h1m->flags & H1_MF_XFER_ENC)) {
+						h1m->flags |= H1_MF_DUPLICATE_TE;
+					}
+					te_header_seen = 1;
 				}
 				else if (isteqi(n, ist("content-length"))) {
 					ret = h1_parse_cont_len_header(h1m, &v);
diff --git a/src/mux_h1.c b/src/mux_h1.c
index 1d8fb3c5f..05e1a1aa5 100644
--- a/src/mux_h1.c
+++ b/src/mux_h1.c
@@ -1499,6 +1499,10 @@ static size_t h1_process_input(struct h1c *h1c, struct buffer *buf, size_t count
 			TRACE_USER((!(h1m->flags & H1_MF_RESP) ? "rcvd H1 request headers" : "rcvd H1 response headers"),
 				   H1_EV_RX_DATA|H1_EV_RX_HDRS, h1c->conn, h1s, htx, (size_t){ret});
 
+			if ((h1m->flags & H1_MF_DUPLICATE_TE) && (h1c->px->cap & PR_CAP_BE)) {
+				h1c->px->be_counters.duplicate_te_header++;
+			}
+
 			if ((h1m->flags & H1_MF_RESP) &&
 			    h1s->status < 200 && (h1s->status == 100 || h1s->status >= 102)) {
 				h1m_init_res(&h1s->res);
diff --git a/src/stats.c b/src/stats.c
index ea023bb57..2707ab8e2 100644
--- a/src/stats.c
+++ b/src/stats.c
@@ -249,6 +249,7 @@ const struct name_desc stat_fieldsST_F_TOTAL_FIELDS = {
 	ST_F_SAFE_CONN_CUR                 = { .name = "safe_conn_cur",               .desc = "Current number of safe idle connections"},
 	ST_F_USED_CONN_CUR                 = { .name = "used_conn_cur",               .desc = "Current number of connections in use"},
 	ST_F_NEED_CONN_EST                 = { .name = "need_conn_est",               .desc = "Estimated needed number of connections"},
+	ST_F_TE_DUP                        = { .name = "te_dup",                      .desc = "Total number of duplicate Transfer-Encoding headers detected" },
 };
 
 /* one line of info */
@@ -1977,6 +1978,8 @@ int stats_fill_be_stats(struct proxy *px, int flags, struct field *stats, int le
 	statsST_F_RT_MAX       = mkf_u32(FN_MAX, px->be_counters.dtime_max);
 	statsST_F_TT_MAX       = mkf_u32(FN_MAX, px->be_counters.ttime_max);
 
+	statsST_F_TE_DUP       = mkf_u64(FN_COUNTER, px->be_counters.duplicate_te_header);
+
 	return 1;
 }
 
-- 
2.44.1
```

**Comment by frobware** [comment]
Reproducer steps/setup in https://github.com/frobware/haproxy-hacks/tree/master/OCPBUGS-40850.

**Comment by frobware** [comment]
Note: these rules may be relaxed in future release of haproxy: 

https://www.mail-archive.com/haproxy@formilux.org/msg45314.html:

>   Some invalid Transfer-Encoding values are now accepted during the H1
    response parsing when accept-invalid-http-response option is enabled,
    even if it is forbidden by the RFC-9112. So, now, with this option,
    multiple "chunked" values are accepted, as well as empty values. When
    several "chunked" values are found, the payload will still be considered
    as encoded once and the header will be sanitized when sent to the
    client. The request parsing was not changed. This remains forbidden
    because it is highly suspicious if a client is sending an invalid T-E
    header. On server side, we can consider the server as trusted. But you
    must still remain careful with such behavior. And, of course, the best
    is to fix the application.

**Comment by frobware** [comment]
Internal Slack discussion: https://redhat-internal.slack.com/archives/CCH60A77E/p1726756870221739?thread_ts=1725986884.497919&cid=CCH60A77E

**Comment by frobware** [comment]
/hold

Still testing.

**Comment by frobware** [comment]
Unit tests [do] need updating: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_router/626/pull-ci-openshift-router-release-4.13-unit/1836840431090929664/build-log.txt.

**Comment by alebedev87** [comment]
@frobware : I see that the flag name for the Transfer-Encoding presence is `H1_MF_XFER_ENC`. Do we want to try to respect similar naming for the duplicate one? Something like: `H1_MF_XFER_ENC_DUP`.

**Comment by frobware** [comment]
> @frobware : I see that the flag name for the Transfer-Encoding presence is `H1_MF_XFER_ENC`. Do we want to try to respect similar naming for the duplicate one? Something like: `H1_MF_XFER_ENC_DUP`.

I can do. Though I'd prefer to gather all the changes we want rather than doing this piecemeal. Anything else in the C change that concerns you?

**Comment by alebedev87** [comment]
> Anything else in the C change that concerns you?

No, the HAProxy change looks fine to me.

I assume you plan to push the HAProxy patch first and then drop the change in `images/router/haproxy/Dockerfile.rhel8` from this PR, right?

**Comment by frobware** [comment]
> > @frobware : I see that the flag name for the Transfer-Encoding presence is `H1_MF_XFER_ENC`. Do we want to try to respect similar naming for the duplicate one? Something like: `H1_MF_XFER_ENC_DUP`.
> 
> I can do. Though I'd prefer to gather all the changes we want rather than doing this piecemeal. Anything else in the C change that concerns you?

@alebedev87 and I discussed offline and decided to keep the current name to maintain compatibility with the existing image builds that customers are using. While the proposed name is better, it doesn't change the image's behaviour, and we don't want to invalidate the testing that has already been done.

**Comment by frobware** [comment]
> > Anything else in the C change that concerns you?
> 
> No, the HAProxy change looks fine to me.
> 
> I assume you plan to push the HAProxy patch first and then drop the change in `images/router/haproxy/Dockerfile.rhel8` from this PR, right?

Yes, to both. 

/hold 

We need to drop the 'DO NOT MERGE' commit. And the haproxy change needs to land first.

**Comment by frobware** [comment]
Patch to be applied to https://pkgs.devel.redhat.com/cgit/rpms/haproxy/log/?h=rhaos-4.13-rhel-8.

```diff
From f2675ee751cb096bb54572206217c5de070c0f02 Mon Sep 17 00:00:00 2001
From: Andrew McDermott <amcdermo@redhat.com>
Date: Mon, 16 Sep 2024 15:31:14 +0100
Subject: Detect duplicate Transfer-Encoding headers (OCPBUGS-48050)

Resolve https://issues.redhat.com/browse/OCPBUGS-40850.

I also raised the following upstream issue[1] which specifically calls
out the patch we created and are now carrying in this commit.

[1] https://github.com/haproxy/haproxy/issues/2721
---
 ...-duplicate-Transfer-Encoding-headers.patch | 118 ++++++++++++++++++
 haproxy.spec                                  |  10 +-
 2 files changed, 127 insertions(+), 1 deletion(-)
 create mode 100644 0001-Record-duplicate-Transfer-Encoding-headers.patch

diff --git a/0001-Record-duplicate-Transfer-Encoding-headers.patch b/0001-Record-duplicate-Transfer-Encoding-headers.patch
new file mode 100644
index 0000000..ee6b97f
--- /dev/null
+++ b/0001-Record-duplicate-Transfer-Encoding-headers.patch
@@ -0,0 +1,118 @@
+From 0b3577ce1e0236498756fba65ee0ef9d436f8a10 Mon Sep 17 00:00:00 2001
+From: Andrew McDermott <amcdermo@redhat.com>
+Date: Thu, 19 Sep 2024 10:56:53 +0100
+Subject: Record duplicate Transfer-Encoding headers
+
+Add support for tracking servers responding with duplicate
+Transfer-Encoding headers within a single response. Increment the
+counter for the relevant backend when duplicates are detected. Introduce
+the new te_dup stat to enable interrogation of backends experiencing
+this issue through the stats socket.
+---
+ include/haproxy/counters-t.h | 2 ++
+ include/haproxy/h1.h         | 1 +
+ include/haproxy/stats-t.h    | 1 +
+ src/h1.c                     | 5 +++++
+ src/mux_h1.c                 | 4 ++++
+ src/stats.c                  | 3 +++
+ 6 files changed, 16 insertions(+)
+
+diff --git a/include/haproxy/counters-t.h b/include/haproxy/counters-t.h
+index b896c4d36..05d2f69aa 100644
+--- a/include/haproxy/counters-t.h
++++ b/include/haproxy/counters-t.h
+@@ -86,6 +86,8 @@ struct be_counters {
+ 	long long denied_req;                   /* blocked requests because of security concerns */
+ 	long long denied_resp;                  /* blocked responses because of security concerns */
+ 
++	long long duplicate_te_header;          /* number of times responses included duplicated Transfer-Encoding header */
++
+ 	long long connect;                      /* number of connection establishment attempts */
+ 	long long reuse;                        /* number of connection reuses */
+ 	long long failed_conns;                 /* failed connect() attempts (BE only) */
+diff --git a/include/haproxy/h1.h b/include/haproxy/h1.h
+index 8835f37f6..f1284f53b 100644
+--- a/include/haproxy/h1.h
++++ b/include/haproxy/h1.h
+@@ -95,6 +95,7 @@ enum h1m_state {
+ #define H1_MF_CLEAN_CONN_HDR    0x00001000 // skip close/keep-alive values of connection headers during parsing
+ #define H1_MF_METH_CONNECT      0x00002000 // Set for a response to a CONNECT request
+ #define H1_MF_METH_HEAD         0x00004000 // Set for a response to a HEAD request
++#define H1_MF_DUPLICATE_TE      0x00008000 // Set when a duplicated Transfer-Encoding header is detected
+ 
+ /* Note: for a connection to be persistent, we need this for the request :
+  *   - one of CLEN or CHNK
+diff --git a/include/haproxy/stats-t.h b/include/haproxy/stats-t.h
+index 300454a14..caa318c09 100644
+--- a/include/haproxy/stats-t.h
++++ b/include/haproxy/stats-t.h
+@@ -429,6 +429,7 @@ enum stat_field {
+ 	ST_F_SAFE_CONN_CUR,
+ 	ST_F_USED_CONN_CUR,
+ 	ST_F_NEED_CONN_EST,
++	ST_F_TE_DUP,
+ 
+ 	/* must always be the last one */
+ 	ST_F_TOTAL_FIELDS
+diff --git a/src/h1.c b/src/h1.c
+index bb8acfb34..07b92fb4c 100644
+--- a/src/h1.c
++++ b/src/h1.c
+@@ -277,6 +277,7 @@ int h1_headers_to_hdr_list(char *start, const char *stop,
+ 	unsigned int col;  /* position of the colon */
+ 	unsigned int eol;  /* end of line */
+ 	unsigned int sov;  /* start of value */
++	unsigned int te_header_seen = 0; /* Tracks if a Transfer-Encoding header has been encountered. */
+ 	union h1_sl sl;
+ 	int skip_update;
+ 	int restarting;
+@@ -807,6 +808,10 @@ int h1_headers_to_hdr_list(char *start, const char *stop,
+ 
+ 				if (isteqi(n, ist("transfer-encoding"))) {
+ 					h1_parse_xfer_enc_header(h1m, v);
++					if (te_header_seen && (h1m->flags & H1_MF_XFER_ENC)) {
++						h1m->flags |= H1_MF_DUPLICATE_TE;
++					}
++					te_header_seen = 1;
+ 				}
+ 				else if (isteqi(n, ist("content-length"))) {
+ 					ret = h1_parse_cont_len_header(h1m, &v);
+diff --git a/src/mux_h1.c b/src/mux_h1.c
+index 1d8fb3c5f..05e1a1aa5 100644
+--- a/src/mux_h1.c
++++ b/src/mux_h1.c
+@@ -1499,6 +1499,10 @@ static size_t h1_process_input(struct h1c *h1c, struct buffer *buf, size_t count
+ 			TRACE_USER((!(h1m->flags & H1_MF_RESP) ? "rcvd H1 request headers" : "rcvd H1 response headers"),
+ 				   H1_EV_RX_DATA|H1_EV_RX_HDRS, h1c->conn, h1s, htx, (size_t[]){ret});
+ 
++			if ((h1m->flags & H1_MF_DUPLICATE_TE) && (h1c->px->cap & PR_CAP_BE)) {
++				h1c->px->be_counters.duplicate_te_header++;
++			}
++
+ 			if ((h1m->flags & H1_MF_RESP) &&
+ 			    h1s->status < 200 && (h1s->status == 100 || h1s->status >= 102)) {
+ 				h1m_init_res(&h1s->res);
+diff --git a/src/stats.c b/src/stats.c
+index ea023bb57..2707ab8e2 100644
+--- a/src/stats.c
++++ b/src/stats.c
+@@ -249,6 +249,7 @@ const struct name_desc stat_fields[ST_F_TOTAL_FIELDS] = {
+ 	[ST_F_SAFE_CONN_CUR]                 = { .name = "safe_conn_cur",               .desc = "Current number of safe idle connections"},
+ 	[ST_F_USED_CONN_CUR]                 = { .name = "used_conn_cur",               .desc = "Current number of connections in use"},
+ 	[ST_F_NEED_CONN_EST]                 = { .name = "need_conn_est",               .desc = "Estimated needed number of connections"},
++	[ST_F_TE_DUP]                        = { .name = "te_dup",                      .desc = "Total number of duplicate Transfer-Encoding headers detected" },
+ };
+ 
+ /* one line of info */
+@@ -1977,6 +1978,8 @@ int stats_fill_be_stats(struct proxy *px, int flags, struct field *stats, int le
+ 	stats[ST_F_RT_MAX]       = mkf_u32(FN_MAX, px->be_counters.dtime_max);
+ 	stats[ST_F_TT_MAX]       = mkf_u32(FN_MAX, px->be_counters.ttime_max);
+ 
++	stats[ST_F_TE_DUP]       = mkf_u64(FN_COUNTER, px->be_counters.duplicate_te_header);
++
+ 	return 1;
+ }
+ 
+-- 
+2.43.5
+
diff --git a/haproxy.spec b/haproxy.spec
index f0a2155..4e76a59 100644
--- a/haproxy.spec
+++ b/haproxy.spec
@@ -10,7 +10,7 @@
 
 Name:           haproxy
 Version:        2.2.24
-Release:        4.rhaos4.13%{?dist}
+Release:        5.rhaos4.13%{?dist}
 Summary:        Do not ship, install or use this, use %{real_name} subpackage instead
 
 License:        GPLv2+
@@ -27,6 +27,9 @@ Patch1:         0001-BUG-CRITICAL-http-properly-reject-empty-http-header-.patch
 # https://issues.redhat.com/browse/OCPBUGS-20325 (CVE-2023-40225)
 Patch2:         0001-BUG-MAJOR-http-reject-any-empty-content-length-heade.patch
 
+# https://issues.redhat.com/browse/OCPBUGS-40850
+Patch3:         0001-Record-duplicate-Transfer-Encoding-headers.patch
+
 BuildRequires:  openssl-devel
 BuildRequires:  pcre-devel
 BuildRequires:  zlib-devel
@@ -60,6 +63,7 @@ availability environments. Indeed, it can:
 %patch0 -p1
 %patch1 -p1
 %patch2 -p1
+%patch3 -p1
 
 %build
 regparm_opts=
@@ -101,6 +105,10 @@ fi
 %{_sbindir}/%{name}
 
 %changelog
+* Mon Sep 16 2024 Andrew McDermott <amcdermo@redhat.com> - 2.2.24-5.rhaos4.13
+- Detect duplicate Transfer-Encoding headers (OCPBUGS-48050)
+- Resolve https://issues.redhat.com/browse/OCPBUGS-40850.
+
 * Fri Sep 29 2023 Andrew McDermott <amcdermo@redhat.com> - 2.2.24-4.rhaos4.13
 - Resolve https://issues.redhat.com/browse/OCPBUGS-20325 (CVE-2023-40225)
 
-- 
2.44.1
```

**Comment by ShudiLi** [comment]
tested it with 4.13.0-0.ci.test-2024-10-12-024602-ci-ln-tptmp02-latest

```
1.
% oc get clusterversion
NAME      VERSION                                                   AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.13.0-0.ci.test-2024-10-12-024602-ci-ln-tptmp02-latest   True        False         147m    Cluster version is 4.13.0-0.ci.test-2024-10-12-024602-ci-ln-tptmp02-latest

2.
% oc -n test get route
NAME          HOST/PORT                                                                     PATH   SERVICES      PORT          TERMINATION   WILDCARD
unsec-apach   unsec-apach-test.apps.ci-ln-tptmp02-72292.origin-ci-int-gce.dev.rhcloud.com          unsec-apach   unsec-apach                 None

3.
% curl -kv --header 'Content-Type: application/json' --data '{"code": 0}' -H "transfer-encoding: chunked" -H "transfer-encoding: chunked" http://unsec-apach-test.apps.ci-ln-tptmp02-72292.origin-ci-int-gce.dev.rhcloud.com
* Host unsec-apach-test.apps.ci-ln-tptmp02-72292.origin-ci-int-gce.dev.rhcloud.com:80 was resolved.
* IPv6: (none)
* IPv4: 35.224.162.162
*   Trying 35.224.162.162:80...
* Connected to unsec-apach-test.apps.ci-ln-tptmp02-72292.origin-ci-int-gce.dev.rhcloud.com (35.224.162.162) port 80
> POST / HTTP/1.1
> Host: unsec-apach-test.apps.ci-ln-tptmp02-72292.origin-ci-int-gce.dev.rhcloud.com
> User-Agent: curl/8.7.1
> Accept: */*
> Content-Type: application/json
> transfer-encoding: chunked
> transfer-encoding: chunked
> 
* upload completely sent off: 21 bytes
< HTTP/1.1 200 OK
< date: Sat, 12 Oct 2024 05:39:52 GMT
< server: Apache/2.4.37 (centos) OpenSSL/1.1.1k
< upgrade: h2c
< connection: Upgrade
< last-modified: Tue, 19 Mar 2024 10:30:40 GMT
< etag: "e-61400f730f400"
< accept-ranges: bytes
< content-length: 14
< content-type: text/html; charset=UTF-8
< set-cookie: fe27bc5d3b8db7d9308afdc84b692496=7f41807f5c19035ab5dfbf7efb41f6a2; path=/; HttpOnly
< 
It is a test!
* Connection #0 to host unsec-apach-test.apps.ci-ln-tptmp02-72292.origin-ci-int-gce.dev.rhcloud.com left intact

4.
% oc -n openshift-ingress get pods
NAME                              READY   STATUS    RESTARTS   AGE
router-default-7967968795-68tx7   1/1     Running   0          155m
router-default-7967968795-lh6jd   1/1     Running   0          155m

5.
% oc -n openshift-ingress rsh router-default-7967968795-68tx7
sh-4.4$ rpm -qa | grep haproxy
haproxy22-2.2.24-5.rhaos4.13.el8.x86_64
sh-4.4$

6.
sh-4.4$ rpm -qp --changelog https://github.com/frobware/haproxy-builds/raw/master/rhaos-4.13-rhel-8/haproxy22-2.2.24-5.rhaos4.13.el8.x86_64.rpm | more
* Mon Sep 16 2024 Andrew McDermott <amcdermo@redhat.com> - 2.2.24-5.rhaos4.13
- Detect duplicate Transfer-Encoding headers (OCPBUGS-48050)
- Resolve https://issues.redhat.com/browse/OCPBUGS-40850.
```

/label qe-approved
thanks

**Comment by alebedev87** [comment]
[The last patch](https://github.com/openshift/router/pull/626#issuecomment-2407449884) LGTM.

**Comment by frobware** [comment]
This PR depends on [haproxy-2.2.24-5.rhaos4.13.el8_x86_64](https://ocp-artifacts.hosts.prod.psi.rdu2.redhat.com/pub/RHOCP/plashets/4.13/stream/el8/latest/x86_64/os/Packages) which will turn up in https://ocp-artifacts.hosts.prod.psi.rdu2.redhat.com/pub/RHOCP/plashets/4.13/stream/el8/latest/x86_64/os/Packages/.

Brew build: https://brewweb.engineering.redhat.com/brew/taskinfo?taskID=65088974

**Comment by frobware** [comment]
```
% brew latest-build rhaos-4.13-rhel-8-candidate haproxy
Build                                     Tag                   Built by
----------------------------------------  --------------------  ----------------
haproxy-2.2.24-5.rhaos4.13.el8            rhaos-4.13-rhel-8-candidate  amcdermo
```

**Comment by frobware** [comment]
Corrected URL for finding RPMs: https://ocp-artifacts.engineering.redhat.com/pub/RHOCP/plashets/4.13/stream/el8/latest/x86_64/os/Packages/.

And [haproxy-2.2.24-5.rhaos4.13.el8](https://ocp-artifacts.engineering.redhat.com/pub/RHOCP/plashets/4.13/stream/el8/latest/x86_64/os/Packages/haproxy-2.2.24-5.rhaos4.13.el8__x86_64__fd431d51/) is now available so:

/hold cancel

**Comment by candita** [comment]
Retesting because it only passed once in latest round of commits.

/test e2e-upgrade

File: pkg/router/metrics/haproxy/haproxy_test.go [hunk 1]
```diff
@@ -22,108 +22,104 @@ func TestExporter_scrape(t *testing.T) {
...
-be_edge_http:openshift-console:downloads,pod:downloads-564948bf9c-vn6lh:downloads:port:10.128.0.30:8080,0,0,0,0,,0,0,0,,0,,0,0,0,0,UP,256,1,0,0,0,802,0,,1,18,1,,0,,2,0,,0,L4OK,,0,0,0,0,0,0,0,,,,,0,0,,,,,-1,,,0,0,0,0,,,,Layer4 check passed,,2,3,4,,,,10.128.0.30:8080,ce739475136fa468d51cfcf5aad91b68,http,,,,,,,,0,0,0,,,0,,0,0,0,0,
-be_edge_http:openshift-console:downloads,pod:downloads-564948bf9c-g7nsm:downloads:port:10.129.5.61:8080,0,0,0,0,,0,0,0,,0,,0,0,0,0,UP,256,1,0,0,0,802,0,,1,18,2,,0,,2,0,,0,L4OK,,0,0,0,0,0,0,0,,,,,0,0,,,,,-1,,,0,0,0,0,,,,Layer4 check passed,,2,3,4,,,,10.129.5.61:8080,450630300ddc04605decdd966ea57de6,http,,,,,,,,0,0,0,,,0,,0,0,0,0,
-be_edge_http:openshift-console:downloads,BACKEND,0,0,0,0,1,0,0,0,0,0,,0,0,0,0,UP,512,2,0,,0,802,0,,1,18,0,,0,,1,0,,0,,,,0,0,0,0,0,0,,,,0,0,0,0,0,0,0,-1,,,0,0,0,0,,,,,,,,,,,,,a663438294fbd72a8e16964e97c8ecde,http,leastconn,,,,,,,0,0,0,0,0,,,0,0,0,0,
+		`public,FRONTEND,,,0,2,20000,162,18770,30715,0,0,0,,,,,OPEN,,,,,,,,,1,2,0,,,,0,0,0,1,,,,0,160,1,0,1,0,,0,1,162,,,0,0,0,0,,,,,,,,,,,,,,,,,,,,,http,,0,1,162,160,0,0,0,,,0,0,,,,,,,,,,,,,
```
Comment (candita) [inline pkg/router/metrics/haproxy/haproxy_test.go:45]: @frobware were the comments that you removed incorrect?

Comment (frobware) [reply pkg/router/metrics/haproxy/haproxy_test.go:45]: Which comments?

Comment (candita) [reply pkg/router/metrics/haproxy/haproxy_test.go:45]: There was one on L44.  It should show up in the diffs.

Comment (candita) [reply pkg/router/metrics/haproxy/haproxy_test.go:45]: Also L64-66, 86, and 106-108.

Comment (frobware) [reply pkg/router/metrics/haproxy/haproxy_test.go:45]: OK, will resurrect.

**Comment by frobware** [comment]
I used the following test code / setup when reproducing and the issue. https://github.com/frobware/haproxy-hacks/tree/master/OCPBUGS-40850.

File: pkg/router/metrics/haproxy/haproxy_test.go (hunk 1 above)
Comment (frobware) [reply pkg/router/metrics/haproxy/haproxy_test.go:46]: Resolved in [this commit](https://github.com/openshift/router/pull/626/commits/fe66dcc8d4c68eea6ad735fa5072e4f5c562ca8b). The GitHub diff didn't display any comment removal, which I missed as I don't use the side-by-side view. It became clearer after reviewing the screenshot you provided.

**Comment by frobware** [comment]
{  failed to create RBAC objects: could not create role "e2e-aws-serial": Timeout: request did not complete within requested timeout - context deadline exceeded}

/test e2e-aws-serial

**Comment by lihongan** [comment]
@candita seems you need add the label by enter `/label backport-risk-assessed`

**Comment by Miciah** [comment]
/label backport-risk-assessed
per https://github.com/openshift/router/pull/626#issuecomment-2427925069.

**Pull request merged**